File: biglm/kneserney/absolute-discount.py
#!/usr/bin/env python

import sys, itertools, collections

d_max = 3
#mincount = [0,1,1,2,2,2,2,2,2] # this is the SRI-LM default
#mincount = [0,1,1,1,1,1,1,1,1] # no pruning
mincount = [0,1,1,1,1,2,2,2,2] # prune singleton 5-grams

def input():
    for line in sys.stdin:
        context, event, scount = line.rstrip().split('\t')
        yield context, event, int(scount)

if __name__ == "__main__":
    countcount = collections.defaultdict(int)
    for line in open(sys.argv[1]):
        c, cc = [int(s) for s in line.rstrip().split('\t')]
        if c <= d_max+1:
            countcount[c] = cc

    y = float(countcount[1])/(countcount[1]+2*countcount[2])

    # Original Kneser-Ney
    d = [0., y]

    # Modified Kneser-Ney
    d = [0.]+[i-(i+1)*y*countcount[i+1]/countcount[i] for i in xrange(1,d_max+1)]

    sys.stderr.write("Kneser-Ney: D=%s\n" % d)

    for context, records in itertools.groupby(input(), lambda record: record[0]):
        records = list(records)
        sumcount = sum(count for _, _, count in records)
        order = len(context.split())+1
        
        for _, event, count in records:
            if count < mincount[order]:
                prob = 0.
            else:
                discount = d[count] if count < len(d) else d[-1]
                prob = (count-discount)/sumcount

            if context:
                print '%s %s\t%s' % (context, event, prob)
            else:
                print '%s\t%s' % (event, prob)
                

File: biglm/kneserney/old/absolute-discount.pl
#!/usr/bin/env perl

# input: counts
# output: discounted probabilities, with backoff weights

$d_max = 3;

open COUNTS, $ARGV[0] || die;
while (<COUNTS>) {
  chomp;
  ($cc, $c) = split(" ",$_,2);
  $countcount{$c} = $cc if ($c <= $d_max+1);
}
close COUNTS;

#@mincount = (0,1,1,2,2,2,2,2,2); # this is the SRI-LM default
@mincount = (0,1,1,1,1,2,2,2,2);

$d[0] = 0;

# Original Kneser-Ney
#$d[1] = $countcount{1}/($countcount{1}+2*$countcount{2});

# Modified Kneser-Ney
$y = $countcount{1}/($countcount{1}+2*$countcount{2});
for ($i=1; $i<=$d_max; $i++) {
    $d[$i] = $i-($i+1)*$y*$countcount{$i+1}/$countcount{$i};
}

print STDERR "Kneser-Ney: D = ", join(" ", @d), "\n";

sub emit_line {
  my $p;
  my $bow = 0;
  for ($i=0; $i<=$#probs; $i++) {
    $c = $probs[$i];
    $c_ceil = $c > $#d ? $#d : $c;
    if ($c < $mincount[$order]) {
      $probs[$i] = 0;
      $bow += $c/$sum;
    } else {
      $probs[$i] = ($c-$d[$c_ceil])/$sum;
      $bow += $d[$c_ceil]/$sum;
    }
  }
  
  for ($i=0; $i<=$#probs; $i++) {
    $x = $events[$i];
    $p = $probs[$i];
    #print "$p\t$prev $x\t$bow\n";
    print "$p\t$prev $x\n";
  }
}

while (<STDIN>) {
  chomp;
  ($count, @fields) = split;
  $order = $#fields+1;
  $context = join(" ", @fields[0..$#fields-1]);
  $event = $fields[$#fields];
  if (defined($prev) && $prev ne $context) {
    emit_line;
    @probs = ();
    @events = ();
    $sum = 0;
    $n1 = 0;
  }
  $sum += $count;
  $n1 += 1;
  $prev = $context;
  push(@probs, $count);
  push(@events, $event);
}

if (defined($prev)) {
  emit_line;
}




